### PR TITLE
Issue 4224 duration bug

### LINKED ISSUE #4224: [Bug]: [24.5rc0] Simulations with `pybamm.Experiment` terminate after 24 h
### PyBaMM Version

24.5rc0

### Python Version

3.9.13

### Describe the bug

Simulations using `pybamm.Experiment` with a voltage cut-off only appear to terminate with a time cut-off after 24 h.

I'm afraid I haven't had time to get into logs to figure out why this is happening!

### Steps to Reproduce

The following simulation should proceed for approximately 100 h = 360000 s, until the specified voltage cut-off. However, it terminates after 24 h.

```python
import pybamm

parameter_values = pybamm.ParameterValues("Chen2020")
model = pybamm.lithium_ion.DFN()
experiment = pybamm.Experiment(
    [
        "Discharge at 0.01C until 2.5 V",
    ]
)

sim = pybamm.Simulation(
    model,
    parameter_values=parameter_values,
    experiment=experiment,
)

sol = sim.solve()
sim.plot()

print(f'Last time step at {sim.solution["Time [s]"].entries[-1] / 3600} h')
```

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by valentinsulzer**
# Description

Fixes #4224

- ~~longer default duration (1000 hours)~~ 
- default duration is 2 / C-rate hours for C-rate steps
- warning if step finishes due to hitting the default duration

Currently working on adding a warning if maximum duration is reached

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by ejfdickinson** [comment]
@valentinsulzer You need to take `abs(value)` in `CRate.default_duration()`, because `value` comes through negative for charge steps.

**Review by brosaplanella — APPROVED** [review]
Looks good to me, let's just add a line to CHANGELOG before merging.

**Comment by ejfdickinson** [comment]
@brosaplanella @valentinsulzer To confirm, this is working as expected for the use cases on which I was testing.

**Comment by valentinsulzer** [comment]
Changing default duration to 1000 hours caused a severe performance degradation, investigating further

**Comment by aabills** [comment]
Thanks! I just tried it and though I'd prefer a longer duration barring a substantial performance hit, in my opinion, the bug is fixed by the warning alone. 

I wonder if 100 + ε hours might be a happy medium?

**Comment by valentinsulzer** [comment]
This PR adds the warning and also fixes the C/100 case which I assume is more common than specifying a small current (backed up by a n=2 sample size from @aabills and @ejfdickinson ). The default duration is 24h and we can reconsider once @MarcBerliner is done with updates to solvers, which will make solve times much less sensitive to things like duration, period, etc

File: pybamm/simulation.py [hunk 1]
```diff
@@ -768,12 +769,19 @@ def solve(
 
                     logs["termination"] = step_solution.termination
                     # Only allow events specified by experiment
-                    if not (
+                    if (
+                        step_termination == "final time"
+                        and step.uses_default_duration is True
```
Comment (kratman) [inline pybamm/simulation.py:774]: We can probably skip the "is True" part here since it is a bool in the definition

File: pybamm/simulation.py [hunk 2]
```diff
@@ -768,12 +769,19 @@ def solve(
...
+                        break
+                    elif not (
                         isinstance(step_solution, pybamm.EmptySolution)
                         or step_termination == "final time"
                         or "[experiment]" in step_termination
                     ):
-                        callbacks.on_experiment_infeasible(logs)
+                        callbacks.on_experiment_infeasible_event(logs)
                         feasible = False
                         break
```
Comment (kratman) [inline pybamm/simulation.py:786]: It is usually good to end the chains in an else of some sort to be clearer for future devs. A quick workaround is to just break up the last elif:
```python
elif condition:
    code()
```
becomes
```python
else:
    if condition:
        code()
```

Not the prettiest, but it is clearer that the chain is done.

File: pybamm/experiment/step/base_step.py [hunk 3]
```diff
@@ -100,8 +102,11 @@ def __init__(
                     f"Input function must return a real number output at t = {t0}"
                 )
 
+        # Record whether the step uses the default duration
+        # This will be used by the experiment to check whether the step is feasible
+        self.uses_default_duration = duration is None
```
Comment (kratman) [inline pybamm/experiment/step/base_step.py:107]: I assume a zero duration should also trigger the default, not just if one is not passed in, correct?

File: pybamm/experiment/step/base_step.py [hunk 4]
```diff
@@ -71,6 +71,8 @@ def __init__(
         description=None,
         direction=None,
     ):
+        self.input_duration = duration
+        self.input_value = value
```
Comment (kratman) [inline pybamm/experiment/step/base_step.py:75]: If these are only used for the copy method, is there a strong reason not to just use the python copy.deepcopy() method? For the array items the constructor is looping over the arrays anyway, so I would assume python's built in copy would be safer, cleaner and potentially faster than writing our own copy operation.

**Review by kratman — CHANGES_REQUESTED** [review]

File: pybamm/experiment/step/base_step.py (hunk 3 above)
Comment (valentinsulzer) [reply pybamm/experiment/step/base_step.py:107]: Zero duration should error, I'll add a separate check for that

File: pybamm/experiment/step/base_step.py (hunk 4 above)
Comment (valentinsulzer) [reply pybamm/experiment/step/base_step.py:75]: In general I've found deepcopy to be a huge performance bottleneck, and there's a few attributes of step that we don't want to copy (start time and end time, which are calculated based on the adjacent steps). The proper fix willl be to refactor steps so we don't have to copy at all, but that is outside the scope of this PR

**Review by kratman — APPROVED** [review]

**Pull request merged**